omniscidb/ArrowStorage/ArrowStorage.cpp

```diff
@@ -630,6 +630,177 @@ TableInfoPtr ArrowStorage::importArrowTable(std::shared_ptr<arrow::Table> at,
+  new_table.col_data = table.col_data;
+  new_table.row_count = table.row_count;
+  new_table.table_stats = table.table_stats;
+  return res;
+}
+
+void ArrowStorage::refragmentTable(const std::string& table_name,
+                                   const size_t new_frag_size) {
+  auto tinfo = getTableInfo(db_id_, table_name);
+  if (!tinfo) {
+    throw std::runtime_error("Unknown table: "s + table_name);
+  }
+  if (!new_frag_size) {
+    throw std::runtime_error("Cannot refragment to fragment size 0");
+  }
+  mapd_shared_lock<mapd_shared_mutex> data_lock(data_mutex_);
+  auto& table = *tables_.at(tinfo->table_id);
+  mapd_unique_lock<mapd_shared_mutex> table_lock(table.mutex);
+  data_lock.unlock();
+  const size_t new_frag_count =
+      table.row_count / new_frag_size + ((table.row_count % new_frag_size) != 0);
+  std::vector<DataFragment> new_fragments(new_frag_count);
+
+  for (size_t new_frag_idx = 0; new_frag_idx < new_frag_count - 1; new_frag_idx++) {
+    new_fragments[new_frag_idx].metadata.resize(table.col_data.size());
+    new_fragments[new_frag_idx].offset = new_frag_idx * new_frag_size;
+    new_fragments[new_frag_idx].row_count = new_frag_size;
+  }
+  new_fragments[new_frag_count - 1].metadata.resize(table.col_data.size());
+  new_fragments[new_frag_count - 1].offset = (new_frag_count - 1) * new_frag_size;
+  new_fragments[new_frag_count - 1].row_count =
+      table.row_count - new_fragments[new_frag_count - 1].offset;
+  std::vector<bool> lazy_fetch_cols(table.col_data.size(), false);
+
+  if (config_->storage.enable_lazy_dict_materialization) {
+    VLOG(1) << "Refragmenting with lazy dictionary materialization enabled";
+    mapd_shared_lock<mapd_shared_mutex> dict_lock(dict_mutex_);
+    for (size_t col_idx = 0; col_idx < lazy_fetch_cols.size(); col_idx++) {
+      auto col_info = getColumnInfo(db_id_, tinfo->table_id, columnId(col_idx));
+      CHECK(col_info);
+      auto col_type = col_info->type;
+      auto col_arr = table.col_data[col_idx];
+      if (col_type->isExtDictionary() && col_arr->type()->id() == arrow::Type::STRING) {
+        auto dict_data =
+            dicts_.at(dynamic_cast<const hdk::ir::ExtDictionaryType*>(col_type)->dictId())
+                .get();
+        CHECK(dict_data);
+        if (!dict_data->is_materialized) {
+          lazy_fetch_cols[col_idx] = true;
+        }
+      }
+    }
+  }
+  tbb::parallel_for(tbb::blocked_range(0, (int)table.col_data.size()), [&](auto range) {
+    for (auto col_idx = range.begin(); col_idx != range.end(); col_idx++) {
+      auto col_info = getColumnInfo(db_id_, tinfo->table_id, columnId(col_idx));
+      auto col_type = col_info->type;
+      auto col_arr = table.col_data[col_idx];
+      auto elem_type =
+          col_type->isArray()
+              ? dynamic_cast<const hdk::ir::ArrayBaseType*>(col_type)->elemType()
+              : col_type;
+      bool compute_stats = !col_type->isString();
+      if (compute_stats) {
```

Comment:
I think it's better to move this condition out of cycle. Earlier you already checked which frag_ids that have String type.